### PR TITLE
Kamu 149 create a new empty dataset

### PR DESCRIPTION
**Pull request by dmitriy-borzenko**

### CODE REVIEW

File: src/app/dataset-create/dataset-create.component.ts [hunk 1]
```diff
@@ -1,9 +1,135 @@
...
+    }
+
+    public onCreateDataset(): void {
+        this.showMonacoEditor
+            ? this.createDatasetFromSnapshot()
+            : this.createDatasetFromForm();
+    }
+
+    public onFileSelected(event: Event): void {
+        this.yamlTemplate = "";
```
Comment (zaychenko-sergei) [inline src/app/dataset-create/dataset-create.component.ts:78]: Maybe clean existing after we checked there is some file?

File: src/app/dataset-create/dataset-create.service.ts [hunk 2]
```diff
@@ -1,4 +1,84 @@
...
-export class AppDatasetCreateService {}
+@Injectable({ providedIn: "root" })
+export class AppDatasetCreateService {
+    private errorMessageChanges$: Subject<string> = new Subject<string>();
+
+    public get onErrorMessageChanges(): Observable<string> {
+        return this.errorMessageChanges$.asObservable();
+    }
+
+    public errorMessageChanges(message: string): void {
```
Comment (zaychenko-sergei) [inline src/app/dataset-create/dataset-create.service.ts:21]: Unlikely public

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]

**Comment by zaychenko-sergei** [comment]
I don't see a proper reaction on structural errors in Yaml. The query fails, message is not displayed.
image

Note: I changed the name of the snapshot document class.

**Comment by zaychenko-sergei** [comment]
Empty created datasets need a better handling in the existing dataset viewer tabs.
In particular, data tab should display a message there is no data schema or data yet, and should not allow sending queries to the void:

image

Metadata tab could look friendlier too and invite user for interaction:

image

**Comment by zaychenko-sergei** [comment]
When I had a Yaml error, and then disabled the checkbox to create from the snapshot, error message should be cleared:

image

**Comment by zaychenko-sergei** [comment]
Of course, don't forget to test this very well, especially error responses.

File: src/app/dataset-create/dataset-create.component.ts [hunk 3]
```diff
@@ -1,9 +1,135 @@
...
     changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class DatasetCreateComponent {}
+export class DatasetCreateComponent extends BaseComponent implements OnInit {
+    public yamlTemplate: MaybeNull<string> = "";
+    public fileName: MaybeNull<string> = "";
+    public showMonacoEditor = false;
+    public errorMessage = "";
+    public initialHint = "# You can edit this file\n";
+    public owners = ["kamu"];
```
[the remark this answers is not in the document]
Comment (dmitriy-borzenko) [reply src/app/dataset-create/dataset-create.component.ts:27]: Note on future.

File: src/app/dataset-create/dataset-create.component.ts [hunk 4]
```diff
@@ -1,9 +1,135 @@
...
     styleUrls: ["./dataset-create.component.sass"],
     changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class DatasetCreateComponent {}
+export class DatasetCreateComponent extends BaseComponent implements OnInit {
+    public yamlTemplate: MaybeNull<string> = "";
+    public fileName: MaybeNull<string> = "";
+    public showMonacoEditor = false;
+    public errorMessage = "";
+    public initialHint = "# You can edit this file\n";
```
[the remark this answers is not in the document]
Comment (dmitriy-borzenko) [reply src/app/dataset-create/dataset-create.component.ts:26]: marked as readonly.

File: src/app/dataset-create/dataset-create.component.ts [hunk 5]
```diff
@@ -1,9 +1,135 @@
...
+            minimap: {
+                enabled: false,
+            },
+        };
+    public createDatasetForm: FormGroup = this.fb.group({
+        owner: ["kamu", [Validators.required]],
+        datasetName: ["", [Validators.required]],
+        kind: ["root", [Validators.required]],
+    });
+    private kindMapper: Record<string, DatasetKind> = {
```
[the remark this answers is not in the document]
Comment (dmitriy-borzenko) [reply src/app/dataset-create/dataset-create.component.ts:42]: ок, marked as readonly.

File: src/app/dataset-create/dataset-create.component.ts (hunk 1 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-create/dataset-create.component.ts:78]: Yes, better after checked

File: src/app/dataset-create/dataset-create.service.ts (hunk 2 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-create/dataset-create.service.ts:21]: Fixed.

File: src/app/dataset-create/dataset-create.component.html [hunk 6]
```diff
@@ -1,47 +1,151 @@
...
+            ></ngx-monaco-editor>
+        </div>
+        <input
+            type="file"
+            class="file-input d-none"
+            onclick="this.value = null"
+            (change)="onFileSelected($event)"
+            #fileUpload
+        />
+        <div class="my-4 d-flex justify-content-end align-items-center">
```
[the remark this answers is not in the document]
Comment (dmitriy-borzenko) [reply src/app/dataset-create/dataset-create.component.html:119]: Fixed

File: src/app/dataset-create/dataset-create.component.ts [hunk 7]
```diff
@@ -1,9 +1,135 @@
...
 
 @Component({
     selector: "app-dataset-create",
     templateUrl: "./dataset-create.component.html",
     styleUrls: ["./dataset-create.component.sass"],
     changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class DatasetCreateComponent {}
+export class DatasetCreateComponent extends BaseComponent implements OnInit {
+    public yamlTemplate: MaybeNull<string> = "";
```
[the remark this answers is not in the document]
Comment (dmitriy-borzenko) [reply src/app/dataset-create/dataset-create.component.ts:22]: Fixed.

**Comment by dmitriy-borzenko** [comment]
> When I had a Yaml error, and then disabled the checkbox to create from the snapshot, error message should be cleared:
> 
> image

Fixed

File: src/app/dataset-create/dataset-create.component.html [hunk 8]
```diff
@@ -1,47 +1,151 @@
...
-                    Use this type if you are improving or building upon existing
-                    datasets.
-                </p>
-                <button class="app-active-button">
-                    New Derivative Dataset
-                </button>
-                <span class="dataset-create-tipText">See an example</span>
+            <span class="mx-2 fs-3 align-self-end pb-2">/</span>
+            <div class="w-50">
+                <label for="dataset-name" class="fs-12">Dataset name</label
```
[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply src/app/dataset-create/dataset-create.component.html:21]: Examples of incorrect handling:

- `my.`
- `a..b`

These are not valid identifiers, but validations allows it, unlike server.

File: src/app/dataset-view/additional-components/data-component/data-component.html [hunk 9]
```diff
@@ -168,13 +168,21 @@ <h2 class="box-title align-items-center pb-3 m-0">Query:</h2>
...
             ></app-dynamic-table>
         </ng-container>
+        <ng-container *ngIf="!sqlErrorMarker || !currentSchema || !currentData">
+            <mat-divider></mat-divider>
+            <h3 class="box-title text-center pt-4 author">
+                The current dataset is empty or need schema. Need
+                <a class="add-data-link" href="#"> add data</a> or
+                <a class="add-schema-link" href="#"> add schema</a>.
+            </h3>
+        </ng-container>
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/data-component/data-component.html:186]: I think we should distinguish these 2 states - no schema and no data.
We cannot add data, when there is no schema.
So if schema is not defined yet, we must go through a wizard to define `SetPollingSource` or `SetTransform`.

The situation when we have a schema, but no data yet, is different. We should invite the user to pull some data or, maybe, configure periodic automatic pull.

File: src/app/dataset-view/additional-components/metadata-component/metadata.component.html [hunk 10]
```diff
@@ -472,5 +471,13 @@
                 </ng-container>
             </div>
         </ng-container>
+        <ng-container
+            *ngIf="
+                !currentState?.metadata?.metadata?.currentSource &&
+                !currentState?.metadata?.metadata?.currentTransform
+            "
+        >
+            <h3 class="box-title text-center">There is no metadata.</h3>
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/metadata-component/metadata.component.html:480]: Please point to the same action as in Data tab - to start defining `SetPollingSource` for root or `SetTransform` for derived.

File: src/app/dataset-create/dataset-create.component.ts [hunk 11]
```diff
@@ -1,9 +1,150 @@
...
+    public errorMessage = "";
+    public owners = ["kamu"];
+    public createDatasetForm: FormGroup = this.fb.group({
+        owner: ["kamu", [Validators.required]],
+        datasetName: [
+            "",
+            [
+                Validators.required,
+                Validators.pattern(
+                    /^([a-zA-Z0-9][a-zA-Z0-9-]*)+(\.[a-zA-Z0-9][a-zA-Z0-9-]*)*$/,
```
Comment (zaychenko-sergei) [inline src/app/dataset-create/dataset-create.component.ts:48]: I'd like to see solid test coverage for this validator.

File: src/app/dataset-create/dataset-create.component.ts [hunk 12]
```diff
@@ -1,9 +1,150 @@
...
+        {
+            theme: "vs",
+            language: "yaml",
+            renderLineHighlight: "none",
+            minimap: {
+                enabled: false,
+            },
+        };
+    public yamlTemplate = "";
+    public fileName: MaybeNull<string> = "";
```
Comment (zaychenko-sergei) [inline src/app/dataset-create/dataset-create.component.ts:37]: I don't think we are using `fileName` now for anything.

Comment (dmitriy-borzenko) [reply src/app/dataset-create/dataset-create.component.ts:37]: Fixed, removed

File: src/app/dataset-create/dataset-create.component.ts (hunk 11 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-create/dataset-create.component.ts:48]: Added new test with validation

**Pull request merged**